packetnetworking/test_cli.py

```diff
@@ -0,0 +1,405 @@
+                    None,  # metadata_file
+                    "http://metadata.packet.net/metadata",  # metadata_url
+                    None,  # operating_system
+                    "packet-networking-test",  # rootfs
+                    None,  # resolvers
+                    3,  # verbose
+                    False,  # quiet
+                ),
+            },
+            id="verbose level 3+ (DEBUG)",
+        ),
+        pytest.param(
+            {
+                "args": default_args + ["--quiet"],
+                "exit_code": 0,
+                "called_with": (
+                    None,  # metadata_file
+                    "http://metadata.packet.net/metadata",  # metadata_url
+                    None,  # operating_system
+                    "packet-networking-test",  # rootfs
+                    None,  # resolvers
+                    0,  # verbose
+                    True,  # quiet
+                ),
+            },
+            id="quiet",
+        ),
+    ],
+)
+def test_cli(test, mockit):
+    runner = CliRunner()
+    with runner.isolated_filesystem():
+        os.mkdir("packet-networking-test")
+        with open("metadata.json", "w") as f:
+            f.write("data")
+
+        with patch("builtins.open", mock_open(read_data="data")) as mocked_open:
+            with mockit(cli.try_run) as mocked_try_run:
+                result = runner.invoke(cli.cli, test["args"])
+    assert_output(test, result)
+    assert result.exit_code == test["exit_code"]
+    if not test.get("called_with"):
+        mocked_open.assert_not_called()
+        mocked_try_run.assert_not_called()
+    else:
+        if test["called_with"][0]:
+            mocked_open.assert_called_with(test["called_with"][0], "r")
+            assert mocked_try_run.call_args_list[0].args[1:] == test["called_with"][1:]
+        else:
+            mocked_open.assert_not_called()
+            mocked_try_run.assert_called_with(*test["called_with"])
+
+
+def test_cli_retries_on_exception(mockit):
+    runner = CliRunner()
+    with runner.isolated_filesystem():
+        os.mkdir("packet-networking-test")
+
+        with mockit(cli.try_run) as mocked_try_run:
+            mocked_try_run.side_effect = Exception("fail")
+            with patch("packetnetworking.cli.log") as mocked_log:
+                with mockit(time.sleep) as mocked_time_sleep:
+                    result = runner.invoke(cli.cli, default_args + ["-n", "2"])
+    # max retries is set to 2
+    # try_run gets called twice
+    assert mocked_try_run.call_count == 2
+    # first retry logs a message and sleeps
+    # second retry fails and does not log a message or sleep
+    assert mocked_log.error.call_count == 1
+    assert mocked_time_sleep.call_count == 1
+    mocked_time_sleep.assert_called_with(4)
+
+    # Exception raised, therefore bad exit code
+    assert result.exit_code != 0
+
+
+def test_try_run_with_url(mockit, metadata):
+    with patch("requests.get") as mocked_requests_get:
```

Comment:
these mocks can all be in just one `with` statement if separated with `,`. Might get very long, but that might be better than all the extra indentation :D.

Reply:
interesting, will look into that.